File: BE/src/main/java/com/a204/batang/domain/auth/service/AuthService.java
package com.a204.batang.domain.auth.service;

import com.a204.batang.domain.auth.dto.request.LoginRequest;
import com.a204.batang.domain.auth.dto.request.RefreshTokenRequest;
import com.a204.batang.domain.auth.dto.request.SendEmailCodeRequest;
import com.a204.batang.domain.auth.dto.request.SignupRequest;
import com.a204.batang.domain.auth.dto.request.VerifyEmailCodeRequest;
import com.a204.batang.domain.auth.dto.request.WithdrawRequest;
import com.a204.batang.domain.auth.dto.response.LoginResponse;
import com.a204.batang.domain.auth.dto.response.MemberInfoResponse;
import com.a204.batang.domain.auth.dto.response.RefreshTokenResponse;
import com.a204.batang.domain.auth.dto.response.SendEmailCodeResponse;
import com.a204.batang.domain.auth.dto.response.SignupResponse;
import com.a204.batang.domain.auth.dto.response.VerifyEmailCodeResponse;

import java.util.UUID;

/**
 * 인증/회원 관련 비즈니스 로직을 정의하는 서비스 인터페이스다.
 */
public interface AuthService {

    /**
     * 이메일 인증 코드를 발송한다.
     *
     * @param request 발송 요청
     * @return 발송 결과
     */
    SendEmailCodeResponse sendEmailCode(SendEmailCodeRequest request);

    /**
     * 이메일 인증 코드를 확인하고 인증 토큰을 발급한다.
     *
     * @param request 확인 요청
     * @return 인증 토큰 응답
     */
    VerifyEmailCodeResponse verifyEmailCode(VerifyEmailCodeRequest request);

    /**
     * 회원가입을 처리하고 토큰을 발급한다.
     *
     * @param request 회원가입 요청
     * @return 회원가입 결과
     */
    SignupResponse signup(SignupRequest request);

    /**
     * 이메일과 비밀번호로 로그인한다.
     *
     * @param request 로그인 요청
     * @return 로그인 결과
     */
    LoginResponse login(LoginRequest request);

    /**
     * Refresh Token으로 새 토큰 쌍을 발급한다.
     *
     * @param request 토큰 갱신 요청
     * @return 새 토큰 응답
     */
    RefreshTokenResponse refresh(RefreshTokenRequest request);

    /**
     * Access Token 기준으로 로그아웃을 처리한다.
     *
     * @param accessToken 현재 Access Token
     */
    void logout(String accessToken);

    /**
     * 현재 인증된 회원 정보를 조회한다.
     *
     * @param userId 회원 ID
     * @return 회원 정보
     */
    MemberInfoResponse getMyInfo(UUID userId);

    /**
     * 회원 탈퇴를 처리한다.
     *
     * @param userId 회원 ID
     * @param accessToken 현재 Access Token
     * @param request 탈퇴 요청
     */
    void withdraw(UUID userId, String accessToken, WithdrawRequest request);
}

File: FE/src/features/auth/services/auth.service.ts
import axios from 'axios'
import { api } from '@/shared/lib/axios'
import type { User } from '@/shared/types'

type AppError = Error & { status?: number }

interface ApiResponse<T> {
  status: number
  message: string
  data: T
}

interface UserInfoResponse {
  userId: string
  email: string
  name: string
  userType: 'DESIGNER' | 'CUSTOMER'
}

interface LoginResponse {
  accessToken: string
  refreshToken: string
  user: UserInfoResponse
}

interface SignupResponse {
  userId: string
  email: string
  name: string
  userType: 'DESIGNER' | 'CUSTOMER'
  accessToken: string
  refreshToken: string
}

interface MemberInfoResponse {
  userId: string
  email: string
  name: string
  userType: 'DESIGNER' | 'CUSTOMER'
}

interface SendEmailCodeResponse {
  email: string
  expiresIn: number
}

interface VerifyEmailCodeResponse {
  verifiedToken: string
  expiresIn: number
}

export interface LoginDto {
  email: string
  password: string
}

export interface RegisterDto {
  email: string
  password: string
  name: string
  userType: 'DESIGNER' | 'CUSTOMER'
  verifiedToken: string
}

export interface LoginResult {
  access_token: string
  refresh_token: string
  user: User
}

export interface SendEmailCodeDto {
  email: string
}

export interface VerifyEmailCodeDto {
  email: string
  code: string
}

export interface WithdrawDto {
  password: string
}

const mapUser = (user: UserInfoResponse | MemberInfoResponse): User => ({
  id: user.userId,
  email: user.email,
  name: user.name,
  user_type: user.userType,
  created_at: new Date().toISOString(),
})

const mapSignupUser = (user: SignupResponse): User => ({
  id: user.userId,
  email: user.email,
  name: user.name,
  user_type: user.userType,
  created_at: new Date().toISOString(),
})

const toErrorMessage = (error: unknown, fallback: string) => {
  if (axios.isAxiosError(error)) {
    const message = error.response?.data?.message
    if (typeof message === 'string' && message.trim().length > 0) {
      return message
    }
  }

  if (error instanceof Error && error.message.trim().length > 0) {
    return error.message
  }

  return fallback
}

const toAppError = (error: unknown, fallback: string): AppError => {
  const nextError = new Error(toErrorMessage(error, fallback)) as AppError

  if (axios.isAxiosError(error)) {
    nextError.status = error.response?.status
  }

  return nextError
}

export const authService = {
  login: async (data: LoginDto): Promise<LoginResult> => {
    try {
      const response = await api.post<ApiResponse<LoginResponse>>('/auth/login', {
        email: data.email.trim(),
        password: data.password,
      })

      return {
        access_token: response.data.data.accessToken,
        refresh_token: response.data.data.refreshToken,
        user: mapUser(response.data.data.user),
      }
    } catch (error) {
      throw toAppError(error, '로그인에 실패했습니다. 입력한 정보를 다시 확인해 주세요.')
    }
  },

  register: async (data: RegisterDto): Promise<LoginResult> => {
    try {
      const response = await api.post<ApiResponse<SignupResponse>>('/auth/signup', {
        email: data.email.trim(),
        password: data.password,
        name: data.name.trim(),
        userType: data.userType,
        verifiedToken: data.verifiedToken,
      })

      return {
        access_token: response.data.data.accessToken,
        refresh_token: response.data.data.refreshToken,
        user: mapSignupUser(response.data.data),
      }
    } catch (error) {
      throw toAppError(error, '회원가입에 실패했습니다. 잠시 후 다시 시도해 주세요.')
    }
  },

  logout: async (): Promise<void> => {
    try {
      await api.post('/auth/logout')
    } catch (error) {
      throw toAppError(error, '로그아웃에 실패했습니다.')
    }
  },

  getMe: async (): Promise<User> => {
    try {
      const response = await api.get<ApiResponse<MemberInfoResponse>>('/auth/me')
      return mapUser(response.data.data)
    } catch (error) {
      throw toAppError(error, '사용자 정보를 불러오지 못했습니다.')
    }
  },

  sendEmailCode: async ({ email }: SendEmailCodeDto): Promise<SendEmailCodeResponse> => {
    try {
      const response = await api.post<ApiResponse<SendEmailCodeResponse>>('/auth/email/send-code', {
        email: email.trim(),
      })
      return response.data.data
    } catch (error) {
      throw toAppError(error, '이메일 인증 코드를 전송하지 못했습니다.')
    }
  },

  verifyEmailCode: async ({ email, code }: VerifyEmailCodeDto): Promise<VerifyEmailCodeResponse> => {
    try {
      const response = await api.post<ApiResponse<VerifyEmailCodeResponse>>('/auth/email/verify-code', {
        email: email.trim(),
        code: code.trim(),
      })
      return response.data.data
    } catch (error) {
      throw toAppError(error, '이메일 인증에 실패했습니다.')
    }
  },

  withdraw: async ({ password }: WithdrawDto): Promise<void> => {
    try {
      await api.delete('/auth/me', {
        data: { password },
      })
    } catch (error) {
      throw toAppError(error, '회원탈퇴에 실패했습니다.')
    }
  },
}
